### ISSUE TITLE
waitForThread sometimes causes EXC_BAD_ACCESS

### ISSUE THREAD

**Issue by kylemcdonald** [openframeworks/openFrameworks#4828]
I left a note here https://github.com/openframeworks/openFrameworks/commit/77f1f15019c4557ed09eb628b250c22dfacaa09e#commitcomment-16106595 but those notes are lost easily and @arturoc encouraged me to add an issue.

The bug I noticed is that sometimes calling thread.join() can fail because the `thread.__t_ == 0`. This means when you are (for example) closing an app and call `thread.waitForThread()` in your `ofBaseApp::exit()`, it will throw `EXC_BAD_ACCESS`.

There's a method called `thread.joinable()` which seems to be meant for handling this case, and since adding `if(thread.joinable()) {...}` to that block I haven't gotten any more crashes. But the crashes were rare (1 out of 10 times) so it might just be chance...

While it's still possible for multiple `join` attempts to create a conflict, I think in general, I think it can't hurt to add `if(thread.joinable()) {...}` (especially because the try / catch doesn't catch EXC_BAD_ACCESS).

In the future, I think the more complete solution would be to make sure that joining the thread requires locking the thread, so only one attempt is made at a time to join the thread.

If I can get a very minimal example of this behavior I will post it, but I hope this is helpful for now.

**Comment by arturoc** [comment]
hey thanks, i've gone ahead and added the if(thread.joinable()) anyway, from reading the docs it seems that this might happen when the thread hasn't been initialized yet?

my concern is that at some point the thread can become not joinable while already having checked that but i don't either want to lock there cause it can enter into a deadlock.

anyway i'm going to keep this open and milestone it for 0.10 to review it in case it's still problematic for other people

**Comment by kylemcdonald** [comment]
Just want to add an update to this, I'm still seeing it. Here is a simple example to reproduce the bug:

```c++
#include "ofMain.h"

class ofApp : public ofBaseApp, ofThread {
public:
    void setup(){
        startThread();
    }
    void exit(){
        if(isThreadRunning()){
            waitForThread();
        }
    }
    void threadedFunction(){
        while(isThreadRunning()) {
            sleep(5);
        }
    }
};

int main() {
    ofSetupOpenGL(1280, 720, OF_WINDOW);
    ofRunApp(new ofApp());
}
```

if it doesn't crash the first time, you might need to run it ten times. and try exiting after different amounts of time (1-10 seconds).

the thing that is most surprising is in the screenshot below. the whole time that waitForThread is being called, the id of the thread is non-zero. but somehow, at the moment the code crashes and the debugger pops up, `*this` is NULL.

<img width="960" alt="screen shot 2016-04-10 at 16 44 12" src="https://cloud.githubusercontent.com/assets/157106/14412992/98578284-ff3b-11e5-9e7c-d36c809f8077.png">

if i add a short `ofSleepMillis(100)` just before `cout` above `thread.join()` then i get slightly different output:

```c++
```
ofSleepMillis(100);
cout << "try: " << thread.joinable() << " / " << thread.get_id() << endl;
thread.join();
```
before: 1 / 0x7000003a6000
after: 1 / 0x7000003a6000
try: 0 / 0x0
```

to me, that looks like the thread is getting joined somewhere else? at the very least there's some kind of timing issue here...

**Comment by ofTheo** [comment]
@arturoc @kylemcdonald - I actually just ran into this issue in the current master with ofxKinect. 

I get a crash on the `if( thread.joinable() )`:
https://github.com/openframeworks/openFrameworks/blob/master/libs/openFrameworks/utils/ofThread.cpp#L109
```
		if (INFINITE_JOIN_TIMEOUT == milliseconds){
			if(thread.joinable()){
				try{
					thread.join();
				}catch(...){}
			}
        }else{
            // Wait for "joinWaitMillis" milliseconds for thread to finish
            std::unique_lock<std::mutex> lck(mutex);
            if(!threadDone && condition.wait_for(lck,std::chrono::milliseconds(milliseconds))==std::cv_status::timeout){
				// unable to completely wait for thread
            }
        }
```

If I change this line:
https://github.com/openframeworks/openFrameworks/blob/master/addons/ofxKinect/src/ofxKinect.cpp#L279

from: `waitForThread(false)` to `waitForThread(false, 5000)` ( or any timeout value ) it doesn't crash. 

The code looks quite different based on specifying a timeout. 
I wonder if it could be that the mutex is needed? in the code below?

**Comment by ofTheo** [comment]
@kylemcdonald I get the crash too with your example. However, changing the `waitForThread` in your example code to `waitForThread(true, 5000)`, I don't see the crash.

**Comment by arturoc** [comment]
i think this is a different problem, seems in some recent change to ofMainLoop i introduced a bug where the exit callback was being notified after the application was already destroyed. for me it only happened when calling explicitly std::exit from the application but might have been happening in some other places.

can you try if my latest commit in master fixes it for you?

**Comment by bakercp** [comment]
(@arturoc just confirming the std::exit bug you noted - saw it happening last week I believe)

**Comment by ofTheo** [comment]
hmm - it doesn't seem to fix it with Kyle's example. 
I pulled in and cleaned and I get the same crash if I run it a couple of times. 

<img width="1656" alt="screen shot 2016-08-25 at 10 02 44 pm" src="https://cloud.githubusercontent.com/assets/144000/17991924/de3bdcbe-6b10-11e6-921f-602d08cc4de2.png">

Also changing it to `waitForThread(true, 5000)` which worked before now freezes the app for 5 seconds before crashing. 


<img width="1050" alt="screen shot 2016-08-25 at 10 08 27 pm" src="https://cloud.githubusercontent.com/assets/144000/17991922/d730c010-6b10-11e6-816e-c5e0f9b617db.png">

So I would say the changes might even make it a bit worse, at least with this test code. 
I can test with the Kinect tomorrow as that seemed to crash every time on exit before.

**Comment by arturoc** [comment]
that's really strange it works for me just fine in linux, might be some implementation difference in osx

**Comment by arturoc** [comment]
can you check if exit is called more than once? i pushed a first version that was doing that but then fixed it. in any case if shouldn't be crashing either but just to understand what's going on

**Comment by ofTheo** [comment]
Exit is being called just once. 
And the ofApp destructor isn't being called before the exit() function.  

With the current master, the` waitForThread();` call is still erring in that place.
But `waitForThread(true, 5000);` still works and produces the following:
```
 exit called at 2073
 deconstructor called at 2077
```
 
With this test code.

```
#include "ofMain.h"

class ofApp : public ofBaseApp, ofThread {
public:
    void setup(){
        startThread();
    }
    
    ~ofApp(){
        cout << " deconstructor called at " << ofGetElapsedTimeMillis() << endl;
    }
    
    void exit(){
        cout << " exit called at " << ofGetElapsedTimeMillis() << endl;
        if(isThreadRunning()){
            waitForThread(true, 5000);
        }
    }
    void threadedFunction(){
        while(isThreadRunning()) {
            sleep(5);
        }
    }
};


//========================================================================
int main( ){
	ofSetupOpenGL(1024,768,OF_WINDOW);			// <-------- setup the GL context

	// this kicks off the running of my app
	// can be OF_WINDOW or OF_FULLSCREEN
	// pass in width and height too:
	ofRunApp(new ofApp());

}
```

**Comment by ofTheo** [comment]
Ahh I think I figured it out. 
`thread.detach();` is called after the threaded function returns. 

If this gets called before thread.join() in waitForThread() then the threaded process is no longer owned by the std::thread object.  

commenting out this line makes the example not crash:
https://github.com/openframeworks/openFrameworks/blob/master/libs/openFrameworks/utils/ofThread.cpp#L173

**Comment by ofTheo** [comment]
Yeah it looks like it is to do with the timing between the ofThread::run function and waitForThread. 

If I add ofSleepMillis above detach it crashes every time: 
```
	ofSleepMillis(500);
	thread.detach();
```

If I do the same above join() it doesn't seem to crash at all.

At least as a way of replicating on linux it might be good to try 
```
	ofSleepMillis(500);
	thread.detach();
```

looking at SO:
http://stackoverflow.com/questions/22699313/why-does-this-simple-threaded-c-program-crash-upon-exit-unless-i-call-thread-j
```
In the destructor of std::thread, std::terminate is called if:

the thread was not joined (with t.join())
and was not detached either (with t.detach())
Thus, you should always either join or detach a thread before the flows of execution reaches the destructor.
```

In our case we are doing both. 
Maybe we do detach in the destructor?  Only if join wasn't called?

**Comment by ofTheo** [comment]
hmm no destructor and tricky because we don't really know if someone is going to call waitForThread or not.

what if we replace `detach()` with 

```
if(thread.joinable()){
				try{
					thread.join();
				}catch(...){}
			}
```

Going to give that a try.

**Comment by ofTheo** [comment]
Okay - think I have a pretty good fix. 

If I add the following at the top of waitForThread it sets a bool which prevents detach getting called.  However if stopThread(); is called or waitForThread is called without trying to join, it will allow the detach to take place.  So this should guarantee that both detach and join aren't called. 

```
//-------------------------------------------------
void ofThread::waitForThread(bool callStopThread, long milliseconds){
	if(!threadDone){

		//only need to do this if we are trying to join
		if (INFINITE_JOIN_TIMEOUT == milliseconds ){
			waitForThreadStarted = true;
		}
```

//-------------------------------------------------
void ofThread::run(){
#ifdef TARGET_ANDROID
	JNIEnv * env;
	jint attachResult = ofGetJavaVMPtr()->AttachCurrentThread(&env,nullptr);
	if(attachResult!=0){
		ofLogWarning() << "couldn't attach new thread to java vm";
	}
#endif

	// user function
    // should loop endlessly.
	try{
		threadedFunction();
	}catch(const std::exception& exc){
		ofLogFatalError("ofThreadErrorLogger::exception") << exc.what();
	}catch(...){
		ofLogFatalError("ofThreadErrorLogger::exception") << "Unknown exception.";
	}
	
	//we don't try and detach the thread if its being joined. 
	if(!waitForThreadStarted){
		thread.detach();
	}
```

**Comment by arturoc** [comment]
can you try this instead, in waitForThread:

```cpp
      if (INFINITE_JOIN_TIMEOUT == milliseconds){
			std::unique_lock<std::mutex> lck(mutex);
			if(!threadDone){
				condition.wait(lck);
			}
        }
```

should do the same but safer and won't introduce anymore state variables that need to be reseted on every state change

**Comment by ofTheo** [comment]
@arturoc - that seems to work well! 

This is what waitForThread looks like with the change ( just to confirm the fix is correct )
 
```
//-------------------------------------------------
void ofThread::waitForThread(bool callStopThread, long milliseconds){
	if(!threadDone){
		// tell thread to stop
        if(callStopThread){
            stopThread();
		}

		// wait for the thread to finish
        if(isCurrentThread()){
			return; // waitForThread should only be called outside thread
		}

        if (INFINITE_JOIN_TIMEOUT == milliseconds){
			std::unique_lock<std::mutex> lck(mutex);
			if(!threadDone){
				condition.wait(lck);
			}
        }else{
            // Wait for "joinWaitMillis" milliseconds for thread to finish
            std::unique_lock<std::mutex> lck(mutex);
            if(!threadDone && condition.wait_for(lck,std::chrono::milliseconds(milliseconds))==std::cv_status::timeout){
				// unable to completely wait for thread
            }
        }
    }
}
```

**Comment by arturoc** [comment]
yeah that should be it, send a PR if you want
